spanner/cloud-client/src/main/java/com/example/spanner/SpannerSample.java

```diff
@@ -1473,9 +1503,371 @@ static void queryWithQueryOptions(DatabaseClient dbClient) {
+    System.out.println("All backups with backup name containing \"users\":");
+    for (Backup backup : instance.listBackups(Options.filter("name:users")).iterateAll()) {
+      System.out.println(backup);
+    }
+
+    // List all backups for databases whose name contains the text 'bank'.
+    System.out.println("All backups for databases with a name containing \"bank\":");
+    for (Backup backup : instance.listBackups(Options.filter("database:bank")).iterateAll()) {
+      System.out.println(backup);
+    }
+
+    // List all backups that expire before a certain time.
+    System.out.println("All backups that expire before 2019-10-18T02:56:53Z:");
+    for (Backup backup :
+        instance
+            .listBackups(Options.filter("expire_time < \"2019-10-18T02:56:53Z\""))
+            .iterateAll()) {
+      System.out.println(backup);
+    }
+
+    // List all backups with size greater than a certain number of bytes.
+    System.out.println("All backups with size greater than 1000 bytes:");
+    for (Backup backup : instance.listBackups(Options.filter("size_bytes > 1000")).iterateAll()) {
+      System.out.println(backup);
+    }
+
+    // List all backups with a create time after a certain timestamp and that are also ready.
+    System.out.println("All databases created after 2019-10-18T02:56:53Z and that are ready:");
+    for (Backup backup :
+        instance
+            .listBackups(Options.filter("create_time >= \"2019-10-18T02:56:53Z\" AND state:READY"))
+            .iterateAll()) {
+      System.out.println(backup);
+    }
+
+    // List backups using pagination.
+    Page<Backup> page = instance.listBackups(Options.pageSize(10));
+    while (true) {
+      for (Backup backup : page.getValues()) {
+        System.out.println(backup);
+      }
+      if (!page.hasNextPage()) {
+        break;
+      }
+      page = page.getNextPage();
+    }
+  }
+  // [END spanner_list_backups]
+
+  // [START spanner_restore_backup]
+  static void restoreBackup(
+      DatabaseAdminClient dbAdminClient,
+      DatabaseId sourceDatabaseId,
+      DatabaseId restoreToDatabase) {
+    // Get the most recent sample backup of the sample database.
+    String backupName =
+        String.format(
+            "%s_%02d",
+            sourceDatabaseId.getDatabase(), LocalDate.now().get(ChronoField.ALIGNED_WEEK_OF_YEAR));
+    BackupId backupId = BackupId.of(sourceDatabaseId.getInstanceId(), backupName);
+    Backup backup = dbAdminClient.newBackupBuilder(backupId).build();
+    // Initiate the request which returns an OperationFuture.
+    System.out.println(String.format(
+        "Restoring backup [%s] to database [%s]...",
+        backup.getId().toString(),
+        restoreToDatabase.toString()));
+    // Try the restore operation in a retry loop, as there is a limit on the number of restore
+    // operations that is allowed to execute simultaneously, and we should retry if we hit this,
+    // limit.
```

Comment:
Are you hitting this in the samples tests because a single instance is shared? We didn't do this for other languages but I think that might be because we're creating and destroying an instance for each test run.

I'm a bit worried about the complexity of this in terms of a code sample since we just want to show users the simplest possible example.

Reply:
Yep, the Java CI env runs two parallel tests (Java 8 and Java 11) on the same instance, so the restore operation would fail for one of them each time. What I could do is remove this error handling here and add it to the test case. That keeps the sample simple. The test will then run the sample repeatedly until it succeeds, fails with a different error than the expected error, or more than X attempts have failed with the 'more than X pending restores' error.

Reply:
I've moved the retry logic away from the sample and into the test case.